src/osmet.rs

```diff
@@ -0,0 +1,969 @@
+
+    // we only hash boot files if there's a potential match with an OSTree object, so we keep a
+    // cache to avoid recomputing it multiple times
+    let mut cached_boot_files_digests: HashMap<u64, Sha256Digest> = HashMap::new();
+
+    // boot files we were able to match up with OSTree objects
+    let mut mapped_boot_files: HashMap<PathBuf, Sha256Digest> = HashMap::new();
+
+    let objects_dir = {
+        let mut root_path = root.mountpoint().to_path_buf();
+        root_path.push("ostree/repo/objects");
+        root_path
+    };
+
+    let mut mappings: Vec<Mapping> = vec![];
+    let mut mapped_file_count = 0;
+    let mut empty_file_count = 0;
+    for entry in WalkDir::new(objects_dir) {
+        let entry = entry?;
+
+        if !entry.file_type().is_file() {
+            continue;
+        }
+
+        if entry.path().extension() != Some("file".as_ref()) {
+            continue;
+        }
+
+        let extents = fiemap_path(entry.path().as_os_str())?;
+        if extents.is_empty() {
+            empty_file_count += 1;
+            continue;
+        }
+
+        let object = object_path_to_checksum(entry.path())
+            .chain_err(|| format!("invalid object path {:?}", entry.path()))?;
+
+        for extent in extents {
+            mappings.push(Mapping {
+                extent,
+                object: object.clone(),
+            });
+        }
+
+        // and check if this matches a boot file
+        let len = entry.metadata()?.len();
+        if let Entry::Occupied(boot_entry) = boot_files.entry(len) {
+            // we can't use Entry::or_insert_with() here because get_path_digest() is fallible
+            let boot_file_digest = match cached_boot_files_digests.entry(len) {
+                Entry::Vacant(e) => e.insert(get_path_digest(boot_entry.get())?),
+                Entry::Occupied(e) => e.into_mut(),
+            };
+            if get_path_digest(entry.path())? == *boot_file_digest {
+                mapped_boot_files.insert(boot_entry.remove(), object.clone());
+            }
+        }
+
+        mapped_file_count += 1;
+    }
+
+    println!(
+        "Total OSTree objects scanned from /root: {} ({} mapped, {} empty)",
+        mapped_file_count + empty_file_count,
+        mapped_file_count,
+        empty_file_count
+    );
+
+    println!(
+        "Total OSTree objects found in /boot: {}",
+        mapped_boot_files.len()
+    );
+
+    // would be cool to detect and report fragmented vs sparse files here too
+    println!("Total /root extents: {}", mappings.len());
+
+    canonicalize(&mut mappings);
+
+    Ok((
+        OsmetPartition {
+            start_offset,
+            end_offset,
+            mappings,
+        },
+        mapped_boot_files,
+    ))
+}
+
+fn prescan_boot_partition(boot: &Mount) -> Result<HashMap<u64, PathBuf>> {
+    let mut files: HashMap<u64, PathBuf> = HashMap::new();
+
+    for entry in WalkDir::new(boot.mountpoint()) {
```

Comment:
Personally I love using rayon for this type of stuff Just Because We Can and it's easy.

Reply:
Yeah, could do this. Though... this is so far from the hot path compared to other parts, not sure if it's worth the added complexity, even if the API is easy.